File: scripts/codigos_graficos_limpio.py
# -*- coding: utf-8 -*-
"""
Created on Tue Apr 18 08:39:48 2023

@author: hp
"""

import pandas as pd
import numpy as np
from siuba import *
from plotnine import *
import seaborn as sns
import matplotlib.pyplot as plt
import plotly.express as px
from pandas.plotting import parallel_coordinates

from plotly.offline import plot # en Jupyter esta línea no es necesaria

#%%

#########################################################################
########################   Lectura de archivos   ########################
#########################################################################

import os

os.chdir("C:\\Users\\hp master\\Documents\\SciData\\Est_mult_23")

iris = pd.read_csv("iris.csv")

#%%

#### Cambiar nombres a las columnas

iris.columns

iris = (iris >> rename(Sepal_Length = "Sepal.Length",
               Sepal_Width = "Sepal.Width",
               Petal_Length = "Petal.Length",
               Petal_Width = "Petal.Width")
)

iris.columns

#%%
#########################################################################
###############   Histogramas y densidades por columna   ################
#########################################################################



#### Histograma absoluto del ancho de sépalo sin distinguir especies

(ggplot(data=iris) +
  geom_histogram(mapping=aes(x="Sepal_Width"),
                 fill="red",
                 color="white")
)

#### Histograma relativo del ancho de sépalo sin distinguir especies
#### incluyendo densidad

(ggplot(data=iris) +
  geom_histogram(mapping=aes(x="Sepal_Width",
                             y="..density.."),
                 fill="red",
                 color="black") +
  geom_density(mapping=aes(x="Sepal_Width"),
               fill="red",
               color="black",
               alpha=0.5)
)

#### Histograma de longitud de sépalo separado por especies

iris_pl_setosa = (iris >> filter(_.Species=="setosa") >> select(_.Petal_Length)).squeeze()
iris_pl_virginica = (iris >> filter(_.Species=="virginica") >> select(_.Petal_Length)).squeeze()
iris_pl_versicolor = (iris >> filter(_.Species=="versicolor") >> select(_.Petal_Length)).squeeze()

plt.hist(iris_pl_setosa,alpha=0.5)
plt.hist(iris_pl_virginica,alpha=0.5)
plt.hist(iris_pl_versicolor,alpha=0.5)

#### Histograma de cada columna sin separar por especies

### Convertir la tabla iris a formato ordenado (tidy o largo)
iris_ordenado = iris >> gather("Medida","Valor",-_["Species"])

(ggplot(data=iris_ordenado) +
      geom_histogram(mapping=aes(x="Valor"),
                     fill="red",
                     color="darkblue",
                     bins=30) +
      labs(title = "Mi histograma",
           subtitle = "Basado en Iris") +
      theme(
        panel_background = element_rect("black"), 
        panel_grid = element_blank(),
      ) +
      facet_wrap("~Medida")
)


#### Densidad de la longitud de pétalo separado por especies

(ggplot(data=iris) +
  geom_density(mapping=aes(x="Petal_Length",
                           fill="Species"),
               color="black",
               alpha=0.5)
)

#### Histograma con densidad de la longitud de pétalo separado por especies
sns.distplot(iris_pl_setosa)
sns.distplot(iris_pl_virginica)
sns.distplot(iris_pl_versicolor)
plt.show()

#%%

#########################################################################
#########################   Nubes de puntos   ###########################
#########################################################################

########### Nube de puntos sin separar en especies

p = (ggplot(data=iris) +
  geom_point(mapping=aes(x="Sepal_Width",y="Sepal_Length"),color="red")
)

p

########### Nube de puntos separando en especies

r = (ggplot(data=iris) +
  geom_point(mapping=aes(x="Sepal_Width",y="Sepal_Length",color="Species")) +
  scale_color_manual(values=["red","orange","blue"])
)

r

########### Nube de puntos sin separar en especies con histogramas 

sns.jointplot(data=iris,x="Sepal_Length",y="Sepal_Width")

########### Nube de puntos separando en especies con densidades

sns.jointplot(data=iris,x="Sepal_Length",y="Sepal_Width",hue="Species")

########### Matriz de nubes de puntos sin información extra

sns.pairplot(iris,hue="Species")

########### Nube de puntos para tres columnas

nube_3d_sin_especies = px.scatter_3d(iris,x="Sepal_Length",y="Sepal_Width",z="Petal_Length")
plot(nube_3d_sin_especies)

nube_3d_con_especies = px.scatter_3d(iris,x="Sepal_Length",y="Sepal_Width",z="Petal_Length",color="Species")
plot(nube_3d_con_especies)

#%%
#########################################################################
###############   Densidades conjuntas de dos columnas   ################
#########################################################################

setosa = iris >> filter(_.Species == "setosa")
virginica = iris >> filter(_.Species == "virginica")

sns.kdeplot(data=setosa, x="Sepal_Width",y="Sepal_Length",
            cmap="Reds",fill=True,thresh=0.05)
sns.kdeplot(data=virginica, x="Sepal_Width",y="Sepal_Length",
            cmap="Blues",fill=True,thresh=0.05)


#########################################################################
#################   Mapa de calor de correlaciones   ####################
#########################################################################

correlaciones = (iris >> select(-_.Species)).corr()

sns.heatmap(data=correlaciones,
            vmax=1,vmin=-1,
            cmap="Reds",
            linecolor="white",linewidth=0.5,
            square=True,
            annot=True,fmt=".2f")

#########################################################################
#######################   Coordenadas paralelas   #######################
#########################################################################

orden = ["Sepal_Length","Petal_Width","Petal_Length","Sepal_Width","Species"]

parallel_coordinates(iris[orden], "Species",colormap="cool")

File: scripts/codigos_graficos_limpio.R
library(tidyverse) # funciones esenciales para análisis de datos
library(rayshader) # graficador especial para mapas en 3D
library(ggExtra) # Para construir nubes de puntos con histogramas y densidades
library(GGally) # Para construir matrices de nubes de puntos y coordenadas paralelas
library(scatterplot3d) # Para construir nubes de puntos con tres columnas
library(reshape) # Para pasar tablas de formato ancho a largo y viceversa


#########################################################################
########################   Lectura de archivos   ########################
#########################################################################

setwd("C:\\Users\\hp master\\Documents\\SciData\\Est_mult_23")
iris <- read.csv("iris.csv")

#########################################################################
###############   Histogramas y densidades por columna   ################
#########################################################################

ggplot(data=iris) +
  geom_histogram(mapping=aes(x=Sepal.Width),
                 fill="red",
                 color="white")

ggplot(data=iris) +
  geom_histogram(mapping=aes(x=Sepal.Width,
                             fill=Species),
                 alpha=0.5,
                 color="white")

ggplot(data=iris) +
  geom_histogram(mapping=aes(x=Sepal.Width,
                             y=stat(density)),
                 fill="red",
                 color="black") +
  geom_density(mapping=aes(x=Sepal.Width),
               fill="red",
               color="black",
               alpha=0.5)

ggplot(data=iris) +
  geom_histogram(mapping=aes(x=Sepal.Width,
                             y=stat(density),
                             fill=Species)) +
  geom_density(mapping=aes(x=Sepal.Width,
                           fill=Species),
               color="black",
               alpha=0.5)

#########################################################################
###############   Densidades conjuntas de dos columnas   ################
#########################################################################

########### Curvas de contorno

iris_setos_versi = iris %>% filter(Species=="setosa" | Species=="versicolor")

m <- ggplot(data=iris_setos_versi) +
      stat_density_2d(mapping=aes(x=Sepal.Width,
                                  y=Sepal.Length,
                                  fill=..level..,
                                  group=Species,
                                  color=Species),
                      geom="polygon",
                      show.legend = FALSE) +
      geom_density_2d(mapping=aes(x=Sepal.Width,
                                   y=Sepal.Length,
                                   color=Species),
                      show.legend=FALSE) +
      ylim(4,8.2) +
      theme(panel.background = element_rect("white"),
            panel.grid = element_line(color="darkgrey",size=0.1))
  
m  

########### Tridimensional  

plot_gg(m)

#########################################################################
#########################   Nubes de puntos   ###########################
#########################################################################

########### Nube de puntos sin separar en especies

p = ggplot(data=iris) +
  geom_point(mapping=aes(x=Sepal.Width,y=Sepal.Length),color="red")

########### Nube de puntos sin separar en especies con histogramas y densidades

ggMarginal(p,type="histogram",fill="red")
ggMarginal(p,type="densigram",fill="red")
ggMarginal(p,fill="red")

########### Nube de puntos separando en especies

r = ggplot(data=iris) +
  geom_point(mapping=aes(x=Sepal.Width,y=Sepal.Length,color=Species))

########### Nube de puntos separando en especies con histogramas y densidades

ggMarginal(r,type="histogram",groupColor=TRUE,groupFill = TRUE,alpha=0.5)
ggMarginal(r,type="densigram",groupFill = TRUE,alpha=0.5)
ggMarginal(r,groupFill = TRUE,alpha=0.5)


########### Matriz de nubes de puntos sin información extra

ggpairs(data=iris,
        mapping = aes(color=Species,alpha=0.01),
        columns=1:4,
        upper = list(continuous = "points")) +
  theme(panel.background = element_rect("#202020"),
        panel.grid = element_line("blue"),
        axis.title = element_blank())

########### Matriz de nubes de puntos con información extra

ggpairs(data=iris,
        mapping = aes(color=Species,alpha=0.01),
        columns=1:4) +
  theme(panel.background = element_rect("#202020"),
        panel.grid = element_line("blue"),
        axis.title = element_blank())

########### Nube de puntos para tres columnas

colores = c("blue","green","red")
colores = colores[as.factor(iris$Species)]
scatterplot3d(iris[,c(2,1,3)],pch=19,color=colores)


#########################################################################
#################   Mapa de calor de correlaciones   ####################
#########################################################################


correlaciones = cor(iris[,1:4])
correlaciones

df = melt(correlaciones)

colnames(df) = c("x","y","correlación")

ggplot(df, aes(x = x, y = y, fill = correlación)) +
  geom_tile(color="white") +
  geom_text(aes(label = round(correlación,2)), color = "white", size = 8) +
  scale_fill_gradient2(low="white",
                       mid="orange", high="red",
                       guide=guide_colorbar(ticks=FALSE,
                                            barheight=5, limits=c(-1,1))) 
  


#########################################################################
#######################   Coordenadas paralelas   #######################
#########################################################################

ggparcoord(data = iris,
           columns = c(1,4,3,2),
           #alphaLines = 0.1,
           boxplot = TRUE,
           groupColumn = "Species",           
           showPoints = TRUE)  +
  scale_color_brewer(palette = "Set2") +
  theme(panel.background = element_rect("#202020"),
        panel.grid = element_blank(),
        axis.title = element_blank())

